src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationPair.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        private partial class CompilationTracker
+        {
+            /// <summary>
+            /// When we're working with compilations, we often have two: a compilation that does not contain generated files
+            /// (which we might need later to run generators again), and one that has the stale generated files that we might
+            /// be able to reuse as well. In those cases we have to do the same transformations to both, and this gives us
+            /// a handy way to do precisely that while not forking compilations twice if there are no generated files anywhere.
+            /// </summary>
+            internal struct CompilationPair
+            {
+                public CompilationPair(Compilation withoutGeneratedDocuments, Compilation withGeneratedDocuments) : this()
+                {
+                    WithoutGeneratedDocuments = withoutGeneratedDocuments;
+                    WithGeneratedDocuments = withGeneratedDocuments;
+                }
+
+                public Compilation WithoutGeneratedDocuments { get; }
```

Comment:
the name kind of implies this is actually creating a new Compilation without generated docs rather than just getting a value (like the rest of our WithXXX methods), same for the below property.

I wonder if these would just be better as methods with full names like GetCompilationWithoutGeneratedDocuments

Reply:
@dibarbet Would like to keep it a property access, but how about "CompilationWithoutGeneratedDocuments"?

Reply:
Yup, that works I think, clearly indicates as a noun

Reply:
Renamed.